meta/src/ast.rs

```diff
@@ -201,10 +211,86 @@ impl Expr {
     }
 }
 
+pub struct ExprTopDownIterator {
```

Comment:
This is a nice way to solve this. I've been lazy with the optimizer implementations and went for hacky, functional-jerk type of implementation.

Reply:
Using the build in iterator trait has a lot of benefits and would definitely prefer to use it where appropriate.